src/types/PasteConverter.cpp

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+#include "inc/PasteConverter.h"
+
+using namespace Microsoft::Console::Utils;
+
+std::wstring PasteConverter::Convert(const std::wstring& wstr, PasteFlags flags)
+{
+    std::wstring converted{ wstr };
+
+    // Convert Windows-space \r\n line-endings to \r (carriage return) line-endings
+    if (flags & PasteFlags::CarriageReturnNewline)
+    {
+        // Some notes on this implementation:
+        //
+        // - std::regex can do this in a single line, but is somewhat
+        //   overkill for a simple search/replace operation (and its
+        //   performance guarantees aren't exactly stellar)
+        // - The STL doesn't have a simple string search/replace method.
+        //   This fact is lamentable.
+        // - This line-ending conversion is intentionally fairly
+        //   conservative, to avoid stripping out lone \n characters
+        //   where they could conceivably be intentional.
+        std::wstring::size_type pos = 0;
+
+        while ((pos = converted.find(L"\r\n", pos)) != std::wstring::npos)
+        {
+            converted.replace(pos, 2, L"\r");
+        }
+    }
+
+    // Bracketed Paste Mode, invented by xterm and implemented in many popular terminal emulators.
+    // See: http://www.xfree86.org/current/ctlseqs.html#Bracketed%20Paste%20Mode
+    if (flags & PasteFlags::Bracketed)
+    {
+        // For security reasons, control characters should be filtered.
+        // Here ASCII control characters will be removed, except HT(0x09), LF(0x0a), CR(0x0d) and DEL(0x7F).
+        converted.erase(std::remove_if(converted.begin(), converted.end(), [](wchar_t c) {

```

Comment:
I try to minimize the copy opertion so I choose `erase` instead of `regex_replace`. Performance-wise, I think this should be enough for most cases.